docs/ops/config.md
```
# pika 端口
port : 9221

# pika是多线程的, 该参数能够配置pika的线程数量, 不建议配置值超过部署服务器的CPU核心数量 
thread-num : 1

# 处理命令用户请求命令线程池的大小
thread-pool-size : 8

# sync 主从同步时候从库执行主库传递过来命令的线程数量
sync-thread-num : 6

# sync 处理线程的任务队列大小, 不建议修改
sync-buffer-size : 10

# Pika日志目录, 用于存放INFO, WARNING, ERROR日志以及用于同步的binlog(write2fine)文件
log-path : ./log/

# Pika数据目录
db-path : ./db/

# Pika 底层单个rocksdb单个memtable的大小, 设置越大写入性能越好但会在buffer刷盘时带来更大的IO负载, 请依据使用场景合理配置
[RocksDb-Tuning-Guide](https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide)
write-buffer-size : 268435456

# Pika 的连接超时时间配置, 单位为秒, 当连接无请求时(进入sleep状态)开始从配置时间倒计时, 当倒计时为0时pika将强行
# 断开该连接, 可以通过合理配置该参数避免可能出现的pika连接数用尽问题, 该参数默认值为60
timeout : 60

# 密码管理员密码, 默认为空, 如果该参数与下方的userpass参数相同(包括同时为空), 则userpass参数将自动失效, 所有用户均为
# 管理员身份不受userblacklist参数的限制
requirepass : password

# 同步验证密码, 用于slave(从库)连接master(主库)请求同步时进行验证, 该参数需要与master(主库)的requirepass一致
masterauth :

# 用户密码, 默认为空, 如果该参数与上方的userpass参数相同(包括同时为空), 则本参数将自动失效, 所有用户均为管理员身份不
# 受userblacklist参数的限制
userpass : userpass
 
# 指令黑名单, 能够限制通过userpass登录的用户, 这些用户将不能使用黑名单中的指令, 指令之间使用","隔开, 默认为空
# 建议将高风险命令配置在该参数中
userblacklist : FLUSHALL, SHUTDOWN, KEYS, CONFIG

# 分为经典模式和分片模式，[classic | sharding]，经典模式中支持多db的配置
instance-mode : classic

# 经典模式下下指定db的数量，使用方式和redis一致
databases : 1

# 和 codis 一起使用时，slot 的数量
default-slot-num : 1024

# 定义一个副本组又多少个从副本，目前支持的配置选项范围[0, 1, 2, 3, 4], 0代表不开启此功能
replication-num : 0

# 定义在返回客户端之前主副本收到多少个从副本的ACK反馈信息。目前可以配置的选项范围[0, ...replicaiton-num]，0代表不开启此功能。
consensus-level : 0

# Pika的dump文件名称前缀, bgsave后生成的文件将以该前缀命名
dump-prefix : backup-
 
# 守护进程模式  [yes | no]
daemonize : yes
 
# slotmigrate  [yes | no], pika3.0.0暂不支持该参数
#slotmigrate : no

# Pika dump目录设置, bgsave后生成的文件将存放在该目录中
dump-path : /data1/pika9001/dump/

# dump目录过期时间, 单位为天, 默认为0即永不过期
dump-expire: 0

# pidfile Path pid文件目录
pidfile : /data1/pika9001/pid/9001.pid
 
# pika最大连接数配置参数
maxclients : 20000
 
# rocks-db的sst文件体积, sst文件是层级的, 文件越小, 速度越快, 合并代价越低, 但文件数量就会超多, 而文件越大, 速度相对变慢, 合并代价大, 但文件数量会很少, 默认是 20M
target-file-size-base : 20971520

# binlog(write2file)文件保留时间, 7天, 最小为1, 超过7天的文件会被自动清理
expire-logs-days : 7
 
# binlog(write2file)文件最大数量, 200个, 最小为10, 超过200个就开始自动清理, 始终保留200个
expire-logs-nums : 200
 
# root用户连接保证数量：2个, 即时Max Connection用完, 该参数也能确保本地（127.0.0.1）有2个连接可以同来登陆pika
root-connection-num : 2
 
# 慢日志记录时间, 单位为微秒, pika的慢日志记录在pika-ERROR.log中, pika没有类似redis slow log的慢日志提取api
slowlog-log-slower-than : 10000

# slave是否是只读状态(yes/no, 1/0)
# slave-read-only : 0

# Pika db 同步路径配置参数
db-sync-path : ./dbsync/

# 该参数能够控制全量同步时的传输速度, 合理配置该参数能够避免网卡被用尽, 该参数范围为1~1024, 意为:1MB/s~1024MB/s，当该参数
# 被配置为小于0或大于1024时, 该参数会被自动配置为1024
db-sync-speed : -1 (1024MB/s)

# 指定网卡
# network-interface : eth1

# 同步参数配置, 适用于从库节点(slave), 该参数格式为ip:port, 例如192.168.1.2:6666, 启动后该示例会自动向192.168.1.2的
# 6666端口发送同步请求
# slaveof : master-ip:master-port

# 配置双主或Hub需要的server id, 不使用双主或Hub请忽略该参数
server-id : 1

# 双主配置, 不使用双主请忽略以下配置
double-master-ip :	双主对端Ip
double-master-port :	双主对端Port
double-master-server-id :	双主对端server id
 
# 自动全量compact, 通过配置的参数每天定时触发一次自动全量compact, 特别适合存在多数据结构大量过期、删除、key名称复用的场景
# 参数格式为:"启动时间(小时)-结束时间(小时)/磁盘空余空间百分比", 例如你需要配置一个每天在凌晨3点~4点之间自动compact的任务
# 同时该任务仅仅在磁盘空余空间不低于30%的时候执行, 那么应配置为:03-04/30, 该参数默认为空
compact-cron : 

# 自动全量compact, 该参与与compact-cron的区别为, compact-cron每天仅在指定时间段执行, 而compact-interval则以配置时间为周期
# 循环执行, 例如你需要配置一个每4小时执行一次的自动compact任务, 同时该任务仅仅在磁盘空余空间不低于30%的时候执行, 那么该参
# 数应配置为:4/30, 该参数默认为空
compact-interval :

# 从库实例权重设置, 仅配合哨兵使用,无其它功能, 权重低的slave会优先选举为主库, 该参数默认为0(不参与选举)
slave-priority : 

# 该参数仅适用于pika跨版本同步时不同版本的binlog能够兼容并成功解析, 该参数可配置为[new | old]
# 当该参数被配置为new时, 该实例仅能作为3.0.0及以上版本pika的从库, 与pika2.3.3~2.3.5不兼容
# 当该参数被配置为old时, 该时候仅能作为2.3.3~2.3.5版本pika的从库, 与pika3.0.0及以上版本不兼容
# 该参数默认值为new, 该参数可在没有配置同步关系的时候通过config set动态调整, 一旦配置了同步关系则不可动态修改
# 需要先执行slaveof no one关闭同步配置, 之后即可通过config set动态修改
identify-binlog-type : new

# 主从同步流量控制的的窗口，主从高延迟情形下可以通过提高该参数提高同步性能。默认值9000最大值90000。
sync-window-size : 9000

# 处理客户端连接请求的最大缓存大小，默认是268435456(256MB)，范围为[64MB, 1GB]，需要注意的是主从的配置需要一致。
# 单条命令超过此buffer大小，服务端会自动关闭与客户端的连接。
max-conn-rbuf-size : 268435456

###################
#Critical Settings#
#    危险参数      #
###################
# write2file文件体积, 默认为100MB, 一旦启动不可修改,  limited in [1K, 2G]
binlog-file-size : 104857600

# 压缩方式[snappy, zlib, lz4, zstd]默认为snappy, 一旦启动不可修改
# 官方发布的二进制提供默认的snaapy的静态连接。如果需要其他压缩方式请自行下载相应静态库并进行编译。
compression : snappy

# 指定后台flush线程数量, 默认为1, 范围为[1, 4]
max-background-flushes : 1

# 指定后台压缩线程数量, 默认为1, 范围为[1, 4]
max-background-compactions : 1

# 指定后台线程数量, 默认为1, 范围为[1, 4]
max-background-jobs : 1


# DB可以使用的打开文件的数量, 默认为5000
max-cache-files : 5000

# pika实例所拥有的rocksdb实例使用的memtable大小上限，如果rocksdb实际使用超过这个数值，下一次写入会造成刷盘
[Rocksdb-Basic-Tuning](https://github.com/facebook/rocksdb/wiki/Setup-Options-and-Basic-Tuning)
max-write-buffer-size : 10737418240

# 限制命令返回数据的大小，应对类似于keys *等命令，返回值过大将内存耗尽。
max-client-response-size : 1073741824

# pika引擎中层级因子, 用于控制每个层级与上一层级总容量的倍数关系, 默认为10(倍), 允许调整为5(倍) 
max-bytes-for-level-multiplier : 10

# 统计对于key的操作次数，对于操作频繁的一部分key做小规模compaction
# max-cache-statistic-keys 为受监控key的数量，配置为0代表关闭此功能
max-cache-statistic-keys : 0

# 如果开启小规模compaction，如果对于key操作次数超过small-compaction-threshold上限，那么对该key进行compaction
small-compaction-threshold : 5000


```
# 数据目录说明

#### db目录
用于存放pika的所有数据文件，包含5个子目录（5大数据类型）它们是：kv，set，zset，hash，list，从pika3.0.0开始，这些数据结构目录为：hashes，lists，sets，strings，zsets
#### log目录
用于存放所有日志文件，包含：一般日志、警告日志、错误日志、同步日志（binlog）、同步日志节点信息文件（mainfest）
#### dump目录
用于存放快照式备份产生的文件
#### pid目录
用于存放pika的pid文件
#### dbsync目录
用于主从全量同步时存放全量同步所需的文件

docs/ops/SetUpDevEnvironment.md
## Pika 开发调试环境搭建
本篇面向开发者，介绍如何使用 CLion 搭建 Pika 的开发调试环境。

### 1. 下载安装 CLion

CLion 是 JetBrains 推出的 C/C++ IDE，支持 Linux、Windows、macOS 等多个平台，支持多种编译器，包括 GCC、Clang 等。
- 下载地址：[https://www.jetbrains.com/clion/download/](https://www.jetbrains.com/clion/download/)

### 2. clone Pika 源码

```bash
$ git clone https://github.com/OpenAtomFoundation/pika.git
```

### 3. 打开 Pika 项目

选择 Open，打开 Pika 并选择 CMake 项目。
![](../images/clion-open-project.png)

### 4. 配置 CLion 

CLion 默认会使用 Ninja 作为构建工具，但 Pika 使用的是 Makefile，所以需要修改 CLion 的配置。
![](../images/clion-cmake-settings.png)

### 5. 编译 Pika

在 Run/Debug Configuration -> Cmake Application 里选中 Pika， 点击 Debug 按钮，开始编译 Pika。编译完成后，看到如下输出，说明编译成功。
```
Pika module 3.5.0
usage: pika [-hv] [-c conf/file]
-h               -- show this help
-c conf/file     -- config file
-v               -- show version
example: ./output/bin/pika -c ./conf/pika.conf
```

![](../images/clion-build.png)

### 6. 运行并调试 Pika

运行 Pika 时，需要设置 Pika 的配置文件。在 Run/Debug Configuration -> Cmake Application -> Pika 里配置 Pika 的配置文件 `-c ../conf/pika.conf`，如下图所示。

![](../images/clion-run-config.png)

配置好后点击 Debug 按钮，开始运行 Pika。运行成功后，可以在 CLion 的控制台看到 Pika 的输出。

```log
.............          ....     .....       .....           .....         
#################      ####     #####      #####           #######
####         #####     ####     #####    #####            #########       
####          #####    ####     #####  #####             ####  #####      
####         #####     ####     ##### #####             ####    #####     
################       ####     ##### #####            ####      #####
####                   ####     #####   #####         #################   
####                   ####     #####    ######      #####         #####  
####                   ####     #####      ######   #####           ##### 
-----------Pika config end----------
W20230921 17:13:42.294183 20438988 pika.cc:188] your 'limit -n ' of 10240 is not enough for Redis to start. pika have successfully reconfig it to 25000
I20230921 17:13:42.294448 20438988 pika.cc:207] Server at: ../conf/pika.conf
I20230921 17:13:42.295454 20438988 net_interfaces.cc:62] got interface lo0
I20230921 17:13:42.295512 20438988 net_interfaces.cc:104] Using Networker Interface: lo0
I20230921 17:13:42.295610 20438988 net_interfaces.cc:148] got ip 127.0.0.1
I20230921 17:13:42.295637 20438988 pika_server.cc:137] host: 127.0.0.1 port: 9221
I20230921 17:13:42.295723 20438988 pika_server.cc:79] Worker queue limit is 20100
W20230921 17:13:42.295754 20438988 pika_server.cc:80] 0.0.0.0
I20230921 17:13:42.296372 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.297199 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.297837 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.298482 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.299120 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.299800 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.300457 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:42.301016 20438988 pika_binlog.cc:80] Binlog: Manifest file not exist, we create a new one.
I20230921 17:13:42.302006 20438988 net_kqueue.cc:23] create kqueue
I20230921 17:13:44.301479 20438988 pika_slot.cc:63] db0 DB Success
I20230921 17:13:44.301918 20438988 net_util.cc:128] TimerTaskThread Starting...
I20230921 17:13:44.302228 20438988 pika_server.cc:204] Pika Server going to start
I20230921 17:13:44.302265 20438988 rsync_server.cc:48] start RsyncServer ...
```

此时，Pika 已经在 CLion 中运行，可以使用 redis-cli 连接 Pika，默认端口 9221，进行调试。

![](../images/clion-debug.png)

进入断点后，说明 Pika 调试模式运行成功。



CHANGELOG.MD
# v3.5.3

## New features

- Pika supports ACL[#2013](https://github.com/OpenAtomFoundation/pika/pull/2013) @[lqxhub](https://github.com/lqxhub)

- Automatically resume service when Codis dashboard coroutine panics[#2349](https://github.com/OpenAtomFoundation/pika/pull/2349)@[chengyu-l](https://github.com/chengyu-l)

- During the full replication process, the slave node of the pika service does not receive read traffic requests.[#2197](https://github.com/OpenAtomFoundation/pika/pull/2197)  @[tedli](https://github.com/tedli)

- Pika cache adds bimap data type.[#2197](https://github.com/OpenAtomFoundation/pika/pull/2197)  @[chejinge](https://github.com/chejinge)

- Delete the remaining Slots in Sharing mode. There is only DB under Pika, and there are multiple DBs under one Pika.[#2251](https://github.com/OpenAtomFoundation/pika/pull/2251) @[Mixficsol](https://github.com/Mixficsol)

- Pika exporter exposes cache-related data collection indicators.[#2318](https://github.com/OpenAtomFoundation/pika/pull/2318) @[dingxiaoshuai](https://github.com/dingxiaoshuai123)

- Pika supports separation of fast and slow commands.[#2162](https://github.com/OpenAtomFoundation/pika/pull/2162) @[dingxiaoshuai](https://github.com/dingxiaoshuai123)

- After pika executes bgsave, retain the unix timepoint.[#2167](https://github.com/OpenAtomFoundation/pika/pull/2167) @[hero-heng](https://github.com/hero-heng)

- Pika supports dynamic configuration of the disable_auto_compations parameter.[#2257](https://github.com/OpenAtomFoundation/pika/pull/2257) @[hero-heng](https://github.com/hero-heng)

- Pika supports Redis Stream.[#1955](https://github.com/OpenAtomFoundation/pika/pull/1955) @[KKorpse](https://github.com/KKorpse)

- Pika supports large key analysis tools[#2195](https://github.com/OpenAtomFoundation/pika/pull/2195) @[sjcsjc123](https://github.com/sjcsjc123)

- Pika supports dynamic adjustment of Pika cache parameters[#2197](https://github.com/OpenAtomFoundation/pika/pull/2197)  @[chejinge](https://github.com/chejinge)

- Updated Pika benchmark tool to support more interface stress tests.[#2222](https://github.com/OpenAtomFoundation/pika/pull/2222)@[wangshao1](https://github.com/wangshao1)

- Pika Operator supports automatic expansion of pika clusters.[#2121](https://github.com/OpenAtomFoundation/pika/pull/2121)@[machinly](https://github.com/machinly/)

- Add the CompactRange command to support compacting keys within a certain range.[#2163](https://github.com/OpenAtomFoundation/pika/pull/2163)@[u6th9d](https://github.com/u6th9d)

- Add small time cost compaction policy.[#2172](https://github.com/OpenAtomFoundation/pika/pull/2172)@[u6th9d](https://github.com/u6th9d)

- Upgrade RocksDB version to v8.7.3.[#2157](https://github.com/OpenAtomFoundation/pika/pull/2157)@[JasirVoriya](https://github.com/JasirVoriya)

- Pika distributed cluster Codis proxy adds new observable indicators.[#2199](https://github.com/OpenAtomFoundation/pika/pull/2199)@[dingxiaoshuai](https://github.com/dingxiaoshuai123)

- Pika distributed cluster supports automatic failover.[#2386](https://github.com/OpenAtomFoundation/pika/pull/2386)@[chengyu-l](https://github.com/chengyu-l)

## bugfix

- Fixed an issue where Pika would accidentally delete dump files during full replication from the node.[#2377](https://github.com/OpenAtomFoundation/pika/pull/2377)@[wangshao1](https://github.com/wangshao1)

- Fixed the processing logic after the slave node receives an abnormal response packet from the master during the master-slave replication process.[#2319](https://github.com/OpenAtomFoundation/pika/pull/2319)@[wangshao1](https://github.com/wangshao1)

- Call disable compaction when pika executes the shutdown command to improve the process exit speed. [#2345](https://github.com/OpenAtomFoundation/pika/pull/2345) @[panlei-coder](https://github.com/panlei-coder)

- Fix the problem of inaccurate Codis-dashboard Redis Memory value.[#2337](https://github.com/OpenAtomFoundation/pika/pull/2337) @[Mixficsol](https://github.com/Mixficsol)

- The INFO command is time-consuming and optimized to reduce the frequency of disk checks. [#2197](https://github.com/OpenAtomFoundation/pika/pull/2197)  @[chejinge](https://github.com/chejinge)

- Fixed the issue where rsync deletes temporary files with incorrect paths and fails to delete them, causing rocksdb to fail to open.[#2186](https://github.com/OpenAtomFoundation/pika/pull/2186)@[wangshao1](https://github.com/wangshao1)

- Fixed the problem that the compact, bgsave, and info keyspace commands did not specify the db name, resulting in some coredump commands.[#2194](https://github.com/OpenAtomFoundation/pika/pull/2194)@[u6th9d](https://github.com/u6th9d)

- Codis dashboard uses info replication instead of info command to search master ip to reduce the performance impact on Pika. [#2198](https://github.com/OpenAtomFoundation/pika/pull/2198)  @[chenbt-hz](https://github.com/chenbt-hz)

- Fix Pika cache to use edge cases to solve the problem of cache and DB data inconsistency in some scenarios.[#2225](https://github.com/OpenAtomFoundation/pika/pull/2225)  @[chejinge](https://github.com/chejinge)

- Fixed the issue where Segmentation fault would be reported when the dump folder is empty.[#2265](https://github.com/OpenAtomFoundation/pika/pull/2265)  @[chenbt-hz](https://github.com/chenbt-hz)

- Fixed the problem that some command caches did not take effect due to flag calculation errors.[#2217](https://github.com/OpenAtomFoundation/pika/pull/2217) @[lqxhub](https://github.com/lqxhub)

- Fixed the problem that in master-slave replication mode, after the master instance flushdb, the slave instance cannot be accessed due to deadlock.[#2249](https://github.com/OpenAtomFoundation/pika/pull/2249)@[ForestLH](https://github.com/ForestLH)

- Fixed the issue where some commands did not judge the return value of RocksDB.[#2187](https://github.com/OpenAtomFoundation/pika/pull/2187)@[callme-taota](https://github.com/callme-taota)


src/config.cc
/*
 * Copyright (c) 2023-present, Qihoo, Inc.  All rights reserved.
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#include <iostream>
#include <vector>

#include "config_parser.h"
#include "config.h"

namespace pikiwidb {

static void EraseQuotes(PString& str) {
  // convert "hello" to  hello
  if (str.size() < 2) {
    return;
  }
  if (str[0] == '"' && str[str.size() - 1] == '"') {
    str.erase(str.begin());
    str.pop_back();
  }
}

extern std::vector<PString> SplitString(const PString& str, char seperator);

PConfig g_config;

PConfig::PConfig() {
  daemonize = false;
  pidfile = "/var/run/pikiwidb.pid";

  ip = "127.0.0.1";
  port = 9221;
  timeout = 0;

  loglevel = "notice";
  logdir = "stdout";

  databases = 16;

  // rdb
  saveseconds = 999999999;
  savechanges = 999999999;
  rdbcompression = true;
  rdbchecksum = true;
  rdbfullname = "./dump.rdb";

  maxclients = 10000;

  // slow log
  slowlogtime = 0;
  slowlogmaxlen = 128;

  hz = 10;

  includefile = "";

  maxmemory = 2 * 1024 * 1024 * 1024UL;
  maxmemorySamples = 5;
  noeviction = true;

  backend = BackEndNone;
  backendPath = "dump";
  backendHz = 10;
}

bool LoadPikiwiDBConfig(const char* cfgFile, PConfig& cfg) {
  ConfigParser parser;
  if (!parser.Load(cfgFile)) {
    return false;
  }

  if (parser.GetData<PString>("daemonize") == "yes") {
    cfg.daemonize = true;
  } else {
    cfg.daemonize = false;
  }

  cfg.pidfile = parser.GetData<PString>("pidfile", cfg.pidfile);

  cfg.ip = parser.GetData<PString>("bind", cfg.ip);
  cfg.port = parser.GetData<unsigned short>("port");
  cfg.timeout = parser.GetData<int>("timeout");

  cfg.loglevel = parser.GetData<PString>("loglevel", cfg.loglevel);
  cfg.logdir = parser.GetData<PString>("logfile", cfg.logdir);
  EraseQuotes(cfg.logdir);
  if (cfg.logdir.empty()) {
    cfg.logdir = "stdout";
  }

  cfg.databases = parser.GetData<int>("databases", cfg.databases);
  cfg.password = parser.GetData<PString>("requirepass");
  EraseQuotes(cfg.password);

  // alias command
  {
    std::vector<PString> alias(SplitString(parser.GetData<PString>("rename-command"), ' '));
    if (alias.size() % 2 == 0) {
      for (auto it(alias.begin()); it != alias.end();) {
        const PString& oldCmd = *(it++);
        const PString& newCmd = *(it++);
        cfg.aliases[oldCmd] = newCmd;
      }
    }
  }

  // load rdb config
  std::vector<PString> saveInfo(SplitString(parser.GetData<PString>("save"), ' '));
  if (!saveInfo.empty() && saveInfo.size() != 2) {
    EraseQuotes(saveInfo[0]);
    if (!(saveInfo.size() == 1 && saveInfo[0].empty())) {
      std::cerr << "bad format save rdb interval, bad string " << parser.GetData<PString>("save") << std::endl;
      return false;
    }
  } else if (!saveInfo.empty()) {
    cfg.saveseconds = std::stoi(saveInfo[0]);
    cfg.savechanges = std::stoi(saveInfo[1]);
  }

  if (cfg.saveseconds == 0) {
    cfg.saveseconds = 999999999;
  }
  if (cfg.savechanges == 0) {
    cfg.savechanges = 999999999;
  }

  cfg.rdbcompression = (parser.GetData<PString>("rdbcompression") == "yes");
  cfg.rdbchecksum = (parser.GetData<PString>("rdbchecksum") == "yes");

  cfg.rdbfullname = parser.GetData<PString>("dir", "./") + parser.GetData<PString>("dbfilename", "dump.rdb");

  cfg.maxclients = parser.GetData<int>("maxclients", 10000);

  cfg.slowlogtime = parser.GetData<int>("slowlog-log-slower-than", 0);
  cfg.slowlogmaxlen = parser.GetData<int>("slowlog-max-len", cfg.slowlogmaxlen);

  cfg.hz = parser.GetData<int>("hz", 10);

  // load master ip port
  std::vector<PString> master(SplitString(parser.GetData<PString>("slaveof"), ' '));
  if (master.size() == 2) {
    cfg.masterIp = std::move(master[0]);
    cfg.masterPort = static_cast<unsigned short>(std::stoi(master[1]));
  }
  cfg.masterauth = parser.GetData<PString>("masterauth");

  // load modules' names
  cfg.modules = parser.GetDataVector("loadmodule");

  cfg.includefile = parser.GetData<PString>("include");  // TODO multi files include

  // lru cache
  cfg.maxmemory = parser.GetData<uint64_t>("maxmemory", 2 * 1024 * 1024 * 1024UL);
  cfg.maxmemorySamples = parser.GetData<int>("maxmemory-samples", 5);
  cfg.noeviction = (parser.GetData<PString>("maxmemory-policy", "noeviction") == "noeviction");

  // io threads
  cfg.io_threads_num = parser.GetData<int>("io-threads", 1);

  // slave threads
  cfg.slave_threads_num = parser.GetData<int>("slave-threads", 1);

  // backend
  cfg.backend = parser.GetData<int>("backend", BackEndNone);
  cfg.backendPath = parser.GetData<PString>("backendpath", cfg.backendPath);
  EraseQuotes(cfg.backendPath);
  cfg.backendHz = parser.GetData<int>("backendhz", 10);

  return cfg.CheckArgs();
}

bool PConfig::CheckArgs() const {
#define RETURN_IF_FAIL(cond)        \
  if (!(cond)) {                    \
    std::cerr << #cond " failed\n"; \
    return false;                   \
  }

  RETURN_IF_FAIL(port > 0);
  RETURN_IF_FAIL(databases > 0);
  RETURN_IF_FAIL(maxclients > 0);
  RETURN_IF_FAIL(hz > 0 && hz < 500);
  RETURN_IF_FAIL(maxmemory >= 512 * 1024 * 1024UL);
  RETURN_IF_FAIL(maxmemorySamples > 0 && maxmemorySamples < 10);
  RETURN_IF_FAIL(io_threads_num > 0 && io_threads_num < 129); // as redis
  RETURN_IF_FAIL(backend >= BackEndNone && backend < BackEndMax);
  RETURN_IF_FAIL(backendHz >= 1 && backendHz <= 50);

#undef RETURN_IF_FAIL

  return true;
}

bool PConfig::CheckPassword(const PString& pwd) const { return password.empty() || password == pwd; }

}  // namespace pikiwidb
